File: lossy/engine/bitcrush.py
"""Time-domain degradation: bitcrusher + sample rate reducer.

Complementary to the spectral processing.  These create amplitude distortion
(staircase waveforms) and aliasing (metallic overtones) rather than the
spectral artifacts produced by STFT-based loss.

Crush    -- reduce amplitude quantization levels (16-bit down to ~4-bit).
             No dithering -- the correlated distortion IS the desired effect.
Decimate -- zero-order hold sample rate reduction with phase accumulator.
             No anti-alias filter -- the aliasing IS the desired effect.
"""

import numpy as np
from numba import njit


def crush_and_decimate(audio, params):
    """Apply bitcrusher and/or sample rate reducer.

    Args:
        audio: mono float64 array
        params: parameter dict

    Returns:
        processed mono float64 array, same length
    """
    g = float(params.get("global_amount", 1.0))
    crush = float(params.get("crush", 0.0)) * g
    decimate = float(params.get("decimate", 0.0)) * g

    if crush <= 0.0 and decimate <= 0.0:
        return audio.copy()

    return _crush_decimate(audio, crush, decimate)


@njit(cache=True)
def _crush_decimate(audio, crush, decimate):
    n = len(audio)
    out = np.empty(n, dtype=np.float64)

    # Bitcrusher: crush 0 -> off, 1 -> extreme (16 down to 4 bits)
    if crush > 0:
        bits = 16.0 - 12.0 * crush
        quant = 2.0 ** (bits - 1.0)
        for i in range(n):
            out[i] = np.floor(audio[i] * quant + 0.5) / quant
    else:
        for i in range(n):
            out[i] = audio[i]

    # Sample rate reducer (zero-order hold with fractional phase accumulator)
    if decimate > 0:
        rate_factor = 1.0 + 31.0 * decimate   # hold for 1..32 samples
        phase = 0.0
        held = 0.0
        for i in range(n):
            phase += 1.0
            if phase >= rate_factor:
                held = out[i]
                phase -= rate_factor
            out[i] = held

    return out

File: rust/crates/lossy-dsp/src/bitcrush.rs
//! Time-domain degradation: bitcrusher + sample rate reducer.
//!
//! Port of `lossy/engine/bitcrush.py`.
//!
//! Crush    -- reduce amplitude quantization levels (16-bit down to ~4-bit).
//! Decimate -- zero-order hold sample rate reduction with phase accumulator.

use crate::params::LossyParams;

/// Apply bitcrusher and/or sample rate reducer.
pub fn crush_and_decimate(audio: &[f64], params: &LossyParams) -> Vec<f64> {
    let g = params.global_amount;
    let crush = params.crush * g;
    let decimate = params.decimate * g;

    if crush <= 0.0 && decimate <= 0.0 {
        return audio.to_vec();
    }

    let mut out = Vec::with_capacity(audio.len());

    // Bitcrusher: crush 0 -> off, 1 -> extreme (16 down to 4 bits)
    if crush > 0.0 {
        let bits = 16.0 - 12.0 * crush;
        let quant = (2.0_f64).powf(bits - 1.0);
        for &s in audio {
            out.push((s * quant + 0.5).floor() / quant);
        }
    } else {
        out.extend_from_slice(audio);
    }

    // Sample rate reducer (zero-order hold with fractional phase accumulator)
    if decimate > 0.0 {
        let rate_factor = 1.0 + 31.0 * decimate; // hold for 1..32 samples
        let mut phase = 0.0_f64;
        let mut held = 0.0_f64;
        for s in out.iter_mut() {
            phase += 1.0;
            if phase >= rate_factor {
                held = *s;
                phase -= rate_factor;
            }
            *s = held;
        }
    }

    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_passthrough_when_zero() {
        let audio: Vec<f64> = (0..100).map(|i| (i as f64 / 100.0).sin()).collect();
        let params = LossyParams::default(); // crush=0, decimate=0
        let out = crush_and_decimate(&audio, &params);
        assert_eq!(out.len(), audio.len());
        for (a, b) in audio.iter().zip(out.iter()) {
            assert!((a - b).abs() < 1e-15);
        }
    }

    #[test]
    fn test_crush_reduces_precision() {
        let audio: Vec<f64> = (0..1000).map(|i| (i as f64 * 0.01).sin() * 0.5).collect();
        let mut params = LossyParams::default();
        params.crush = 1.0; // extreme crush
        let out = crush_and_decimate(&audio, &params);
        assert_eq!(out.len(), audio.len());
        // With extreme crush, many values should be quantized to the same level
        let unique_values: std::collections::HashSet<u64> =
            out.iter().map(|x| x.to_bits()).collect();
        assert!(unique_values.len() < audio.len());
    }

    #[test]
    fn test_decimate_holds_values() {
        let audio: Vec<f64> = (0..100).map(|i| i as f64).collect();
        let mut params = LossyParams::default();
        params.decimate = 1.0; // extreme decimation (hold for 32 samples)
        let out = crush_and_decimate(&audio, &params);
        // Consecutive samples should be identical (held)
        let mut run_count = 0;
        for i in 1..out.len() {
            if out[i] == out[i - 1] {
                run_count += 1;
            }
        }
        assert!(run_count > 50); // most samples should be held
    }
}
